[WIP][Grad Compression] Unittest to verify allreduce_hook parity

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50851 [WIP][Grad Compression] Unittest to verify allreduce_hook parity**

Improves upon the previous unittest to ensure allreduce_hook results in the same gradients as vanilla allreduce in DDP.

Differential Revision: [D25963654](https://our.internmc.facebook.com/intern/diff/D25963654/)

rohan-varma committed:
[Grad Compression] Unittest to verify allreduce_hook parity

Improves upon the previous unittest to ensure allreduce_hook results in the same gradients as vanilla allreduce in DDP.

Note that this diff removes the powerSGD hook test, will add that in the next PR as it needs to be slightly different (gradients will no longer be equal)

Differential Revision: [D25963654](https://our.internmc.facebook.com/intern/diff/D25963654/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][Grad Compression] Unittest to verify allreduce_hook parity"


Improves upon the previous unittest to ensure allreduce_hook results in the same gradients as vanilla allreduce in DDP.

Differential Revision: [D25963654](https://our.internmc.facebook.com/intern/diff/D25963654/)

[ghstack-poisoned]

SciPioneer commented:
Since this test is only for allreduce_hook, it will be better to add it rather than replace the test for PowerSGD_hook.

rohan-varma committed:
Update on "[WIP][Grad Compression] Unittest to verify allreduce_hook parity"


Improves upon the previous unittest to ensure allreduce_hook results in the same gradients as vanilla allreduce in DDP.

Differential Revision: [D25963654](https://our.internmc.facebook.com/intern/diff/D25963654/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP][Grad Compression] Unittest to verify allreduce_hook parity"


Improves upon the previous unittest to ensure allreduce_hook results in the same gradients as vanilla allreduce in DDP.

Differential Revision: [D25963654](https://our.internmc.facebook.com/intern/diff/D25963654/)

[ghstack-poisoned]
